claudette/core.py

```diff
@@ -274,6 +276,32 @@ def structured(self:Client,
     tcs = [call_func(o.name, o.input, ns=ns) for o in cts if isinstance(o,ToolUseBlock)]
     return tcs
 
+# %% ../00_core.ipynb
+def is_builtin(tp: type):
+    "Returns True for built in primitive types or containers"
+    return (tp in (str, int, float, bool, complex) or tp is None
+        or getattr(tp, '__origin__', None) is not None)  # Pass through all container types
+
+
+def convert(val: Dict, # dictionary argument being passed in
+            tp: type): # type of the tool function input
+    "Convert converts a single argument"
+    if val is None or is_builtin(tp) or not isinstance(val, dict):
+        return val
+    return tp(**val)
+
+# %% ../00_core.ipynb
+def tool(func):
+    hints = get_type_hints(func)
+    @wraps(func)
+    def wrapper(*args, **kwargs):
+        new_args = [convert(arg, hints[p])
+                   for p,arg in zip(inspect.signature(func).parameters, args)]
```

Comment:
```suggestion
        new_args = [convert(arg, hints[p]) for p,arg in zip(inspect.signature(func).parameters, args)]
```

Generally we assume a width of ~150 chars.